oap-server/server-storage-plugin/storage-elasticsearch-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/elasticsearch/base/IndexStructures.java

```diff
@@ -0,0 +1,157 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package org.apache.skywalking.oap.server.storage.plugin.elasticsearch.base;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.stream.Collectors;
+import lombok.Getter;
+import org.apache.skywalking.oap.server.library.client.elasticsearch.ElasticSearchClient;
+
+public class IndexStructures {
+    private final Map<String, Fields> structures;
+    @Getter
+    private final PropertiesExtractor extractor;
+    @Getter
+    private final PropertiesWrapper wrapper;
+
+    public IndexStructures() {
+        this.structures = new HashMap<>();
+        this.extractor = doGetPropertiesExtractor();
+        this.wrapper = doGetPropertiesWrapper();
+    }
+
+    protected PropertiesExtractor doGetPropertiesExtractor() {
+        return mapping -> (Map<String, Object>) ((Map<String, Object>) mapping.get(
+            ElasticSearchClient.TYPE)).get("properties");
+    }
+
+    protected PropertiesWrapper doGetPropertiesWrapper() {
+        return properties -> {
+            HashMap<String, Object> mappings = new HashMap<>();
+            HashMap<String, Object> types = new HashMap<>();
+            mappings.put(ElasticSearchClient.TYPE, types);
+            types.put("properties", properties);
+            return mappings;
+        };
+    }
+
+    public Map<String, Object> getMapping(String tableName) {
+        return wrapper.wrapper(
+            structures.containsKey(tableName) ? structures.get(tableName).properties : new HashMap<>());
+    }
+
+    /**
+     * Add or append field when the current structures don't contain the input structure or having new fields in it.
+     */
+    public void resolveStructure(String tableName, Map<String, Object> mapping) {
```

Comment:
I think this method should be `idempotent`, could we use `structures.put(tableName, fields);` every time?

Reply:
`resolveStructure` could be called multiple times, such as `no-init` mode waiting, like I mentioned in https://github.com/apache/skywalking/pull/6499#discussion_r590203776

Reply:
renamed the function